services/ml/compilation_delegate_dml.cc

```diff
@@ -43,31 +44,43 @@ HRESULT InitializeDirect3D12(ComPtr<ID3D12Device>& d3D12_device,
   }
 #endif
 
-  Microsoft::WRL::ComPtr<IDXGIFactory4> dxgi_factory;
+  Microsoft::WRL::ComPtr<IDXGIFactory6> dxgi_factory;
   HRESULT hr = CreateDXGIFactory1(IID_PPV_ARGS(&dxgi_factory));
   if (FAILED(hr)) {
     LOG(ERROR) << "Failed creating DXGI factory.";
     return hr;
   }
 
   ComPtr<IDXGIAdapter> dxgi_adapter;
-  size_t adapter_index = 0;
-  do {
-    dxgi_adapter = nullptr;
-    hr = dxgi_factory->EnumAdapters(adapter_index, &dxgi_adapter);
-    if (FAILED(hr))
-      return hr;
-    ++adapter_index;
+  size_t adapter_index =
+      preference == mojom::PREFER_LOW_POWER
+          ? 0
+          : preference == mojom::PREFER_SUSTAINED_SPEED ? 1 : 2;
+  // When DXGI_GPU_PREFERENCE_MINIMUM_POWER is specified for the GpuPreference
+  // parameter, the order of preference for the adapter returned in ppvAdapter
+  // will be:
+  // 1. iGPUs (integrated GPUs)
+  // 2. dGPUs (discrete GPUs)
+  // 3. xGPUs (external GPUs)
+  hr = dxgi_factory->EnumAdapterByGpuPreference(
+      adapter_index, DXGI_GPU_PREFERENCE_MINIMUM_POWER,
```

Comment:
How about we map `PREFER_LOW_POWER` to `DXGI_GPU_PREFERENCE_MINIMUM_POWER` and `PREFER_SUSTAINED_SPEED` to `DXGI_GPU_PREFERENCE_HIGH_PERFORMANCE`? We can select the first (index 0) for each preference. So the code can work if there is only iGPU or dGPU.

Reply:
Do you mean both sustianed and low prefence can work when there is only iGPU or dGPU? does it confuse?

Reply:
I understand the preference is just a developer hint. Browser just tries its best to meet that. So if the system has both iGPU and dGPU, `lower power` and `sustained speed` should select right type of GPU. However, if the system only has one GPU, developer's code should still work. This is how WebGL power preference works today. Does it make sense?

Reply:
Thanks for explanation, it looks good to me. PTAL again.